### DISCUSSION TITLE
Human user programatic login for end2end testing scenarious

### DISCUSSION THREAD

**Discussion by davidkalosi** [zitadel/zitadel#7530] [Q&A]
Hi,

I would like to get some advice - for end2end testing purposes I need to somewhat generate human user access tokens in order to test the API interactions. Currently based on reading through the entire documentation this is not possible, only service users can support this via PAT or keys. We have a role to scope mapping logic in the token actions which does not work for service users (see bug https://github.com/zitadel/zitadel/issues/7428).  So I am really stuck right now. 

Previously in keycloak we were using the resource owner grant scheme for the local instance to be able to programatically create users and get their tokens in the end2end testing flows. I am aware of all the potential security implications for the resource owner grant type but enabling it at least it development mode would make the implementation of end2end testing scenarios much easier.

**Comment by dakshitha** [comment]
Hi! You can generate human tokens (get user to log in and get token) and call your APIs or ZITADEL APIs. Take a look at the following examples tested in Postman. 

Test user login flow in web app, get access token and call the ZITADEL user info endpoint - https://zitadel.com/blog/testing-login-with-postman

Test token introspection in ZITADEL with token obtained from user login - https://zitadel.com/blog/testing-token-introspection-with-postman

Do let me know if I interpreted your question incorrectly.

**Reply by davidkalosi** [reply]
almost, need to achieve this without entering the login details. I know there is a session API that allows me to log in a user, what is unclear to me is weather or how I can use that session to get an access token.

**Reply by dakshitha** [reply]
Yes, the easiest way to do it then is by using service users. We are investigating the bug you've reported. @hifabienne fyi.

**Reply by davidkalosi** [reply]
Back to my original question - is it possible to get a human user token with the combinations of session API and OIDC lifecycle API that does not require end user interaction? 

thanks

**Reply by dakshitha** [reply]
At the moment, if you want to obtain OIDC tokens, you need to complete the complete redirect cycle. So a browser needs to be redirected with an auth code to the application and the application calls the token endpoint to obtain the token. We don't have a shortcut that allows you to obtain OIDC tokens directly from a session token.

**Reply by davidkalosi** [reply]
So I was able to implement this programatically. Here is a step by step overview, can also send a go code snippet if interested, adding this to the documentation would provide a bit more clarity. 

Here is the process 

1. send a request to the `/oauth/v2/authorization` endpoint.  Use the PAT of the service user in the `Authorization` header and make sure you set the `x-zitadel-login-client` header with the user id of the service account. Configure your HTTP client NOT to follow redirect and parse out the auth request ID from the `Location` response header. 
2. create a session using the Session API with the username of the user you want to log in 
3. update the session with the password using the session ID and session token from step 2
4. use the OIDC Session API to create a callback URL, use the request ID from step 1  and the session id and token from step 3
5. send a request to the `/oauth/v2/token` endpoint, use the code challenge from step 1 and the authorization code from step 4.

**Reply by dakshitha** [reply]
Awesome @davidkalosi. Glad it worked for you. Yes, could you please add your code here?
